src/circuit_writer/writer.rs

```diff
@@ -109,6 +113,308 @@ impl Ord for AnnotatedCell {
     }
 }
 
+/// This records the steps to access a child element of a struct or an array variable.
+/// These steps are used to navigate through a mutable variable in the scope in order to update its value.
+/// The access pattern can narrow down to a single field, struct or array. For example:
+/// `houses[1].rooms[2].room_size`
+///
+/// The access to the field called "room_size" at the end can be represented as:
+/// Access {
+///   var_name: "houses", // the name of the variable in the current scope
+///   steps: [ Array(1), Field("rooms"), Array(2), Field("room_size") ],
+///   expr: ... // represent the expression that leads to the value of "room_size"
+/// }
+///
+/// where the "Array" / "Field" enum represents the `AccessKind` of a step.
+/// 
+/// This Access type is currently constructed from:
+/// - Expr::Variable
+/// - Expr::FieldAccess
+/// - Expr::ArrayAccess
+/// 
+/// where Expr::Variable initializes the Access with an empty steps.
+/// Expr::FieldAccess or Expr::ArrayAccess constructs the steps.
+/// 
+/// Then the Expr::Assignment uses the Access to trace and update the value of the variable in the scope.
+#[derive(Debug, Clone)]
+pub struct Access<F, V>
+where
+    F: BackendField,
+    V: BackendVar,
+{
+    pub var_name: String,
+    pub steps: Vec<AccessKind>,
+    // the current access node
+    pub expr: Box<ComputedExpr<F, V>>,
+}
+
+impl<F: BackendField, V: BackendVar> Access<F, V> {
+    pub fn new(var_name: &str, steps: &[AccessKind], expr: Box<ComputedExpr<F, V>>) -> Self {
+        Self {
+            var_name: var_name.to_string(),
+            steps: steps.to_vec(),
+            expr,
+        }
+    }
+}
+
+impl<F: BackendField, V: BackendVar> Display for Access<F, V> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let mut s = self.var_name.clone();
+        for step in &self.steps {
+            match step {
+                AccessKind::Field(field) => s.push_str(&format!(".{}", field)),
+                AccessKind::Array(index) => s.push_str(&format!("[{}]", index)),
+            }
+        }
```

Comment:
can't you just write `self.steps.map(ToString::to_string).join()` if you have `Display` implemented on `AccessKind`?